### PR TITLE
hide project actions from non owners

### PR DESCRIPTION
**Pull request by vijayrathore8492**
Signed-off-by: Vijay Kumar Rathore <professional.vijay8492@gmail.com>

## Change Summary

Project EDIT and DELETE buttons are shown to all the users on the project List string.
This PR hides actions from noneligible users.

## Change type

- [ ] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Provide summary of changes.

## Additional information / screenshots (optional)

Anything for maintainers to be made aware of
User 1
Screenshot 2022-11-01 at 19 05 50

User 2
Screenshot 2022-11-01 at 19 06 20

### CODE REVIEW

**Review by wingkwong — COMMENTED** [review]
Is it still WIP? Didn't see the logic to hide project actions. Just FYI, `Api2.ts` is going to be removed soon.

**Comment by vijayrathore8492** [comment]
> Is it still WIP? Didn't see the logic to hide project actions. Just FYI, `Api2.ts` is going to be removed soon.

Added the missing code now.

File: packages/nocodb-sdk/src/lib/Api2.ts [hunk 1]
```diff
@@ -234,12 +235,12 @@ export interface ColumnType {
   visible?: boolean;
   order?: number;
   colOptions?:
-    | LinkToAnotherRecordType
```
Comment (vijayrathore8492) [inline packages/nocodb-sdk/src/lib/Api2.ts:237]: auto save changes from IDE

File: packages/nocodb-sdk/src/lib/Api2.ts [hunk 2]
```diff
@@ -541,83 +542,83 @@ export interface HookLogType {
 
 export type ColumnReqType =
   | {
-      uidt?:
```
Comment (vijayrathore8492) [inline packages/nocodb-sdk/src/lib/Api2.ts:544]: Autosave changes

File: packages/nocodb-sdk/src/lib/Api2.ts [hunk 3]
```diff
@@ -750,8 +751,8 @@ export class HttpClient<SecurityDataType = unknown> {
           property instanceof Blob
             ? property
             : typeof property === 'object' && property !== null
-            ? JSON.stringify(property)
```
Comment (vijayrathore8492) [inline packages/nocodb-sdk/src/lib/Api2.ts:753]: autosave changes

File: packages/nocodb-sdk/src/lib/Api2.ts [hunk 4]
```diff
@@ -809,7 +810,7 @@ export class HttpClient<SecurityDataType = unknown> {
     baseUrl,
     cancelToken,
     ...params
-  }: FullRequestParams & {wrapped?:any}): Promise<T> => {
```
Comment (vijayrathore8492) [inline packages/nocodb-sdk/src/lib/Api2.ts:812]: autosave again

File: packages/nocodb-sdk/src/lib/Api2.ts [hunk 5]
```diff
@@ -821,8 +822,7 @@ export class HttpClient<SecurityDataType = unknown> {
     const responseFormat = format || requestParams.format;
 
     return this.customFetch(
-      `${baseUrl || this.baseUrl || ''}${path}${
```
Comment (vijayrathore8492) [inline packages/nocodb-sdk/src/lib/Api2.ts:824]: autosave

File: packages/nocodb-sdk/src/lib/Api2.ts [hunk 6]
```diff
@@ -846,18 +846,18 @@ export class HttpClient<SecurityDataType = unknown> {
       const data = !responseFormat
         ? r
         : await response[responseFormat]()
-            .then((data) => {
```
Comment (vijayrathore8492) [inline packages/nocodb-sdk/src/lib/Api2.ts:849]: autosave

File: packages/nocodb/src/lib/models/ProjectUser.ts [hunk 7]
```diff
@@ -183,4 +183,8 @@ export default class ProjectUser {
       project_id: projectId,
     });
   }
+
+  static async getAllProjectsUsers(ncMeta = Noco.ncMeta) {
+    return await ncMeta.metaList2(null, null, MetaTable.PROJECT_USERS, {});
+  }
```
Comment (wingkwong) [inline packages/nocodb/src/lib/models/ProjectUser.ts:189]: 1. I think it's better to call `list` (aligned with other models).
2. We can apply cache here. See the below `list` for reference.
https://github.com/nocodb/nocodb/blob/08d73677dc6e794babe930edd8b54e2dc465434f/packages/nocodb/src/lib/models/Plugin.ts#L44-L51
3. I saw there's a function called `getUsersList`. Can we somehow use it?

File: packages/nocodb/src/lib/models/Project.ts [hunk 8]
```diff
@@ -394,4 +398,18 @@ export default class Project implements ProjectType {
...
+    return projectList.map((project) => {
+      return {
+        ...project,
+        roles: projectsUsers.find(
+          (projectUser) =>
+            projectUser.project_id === project.id &&
+            projectUser.fk_user_id === userId
+        )?.roles,
+      };
+    });
```
Comment (wingkwong) [inline packages/nocodb/src/lib/models/Project.ts:413]: We can make it cleaner

```js
    return projectList.map((project) => ({
      ...project,
      roles: projectsUsers.find(
        (projectUser) =>
          projectUser.project_id === project.id &&
          projectUser.fk_user_id === userId
      )?.roles,
    }));
```

**Comment by wingkwong** [comment]
cc: @pranavxc for the role changes

File: packages/nocodb/src/lib/models/ProjectUser.ts (hunk 7 above)
Comment (vijayrathore8492) [reply packages/nocodb/src/lib/models/ProjectUser.ts:189]: doing the change for points 1 and 2.
For point 3 `getUsersList` finds users of a given project and not of all.
I did not want to query for each project so chose to get a list of projectUsers.

**Comment by vijayrathore8492** [comment]
@o1lab Is it part of next release?

**Comment by pranavxc** [comment]
@vijayrathore8492  We will review it and if all is good will include it as part of the next release :+1:

**Comment by pranavxc** [comment]
Hi @vijayrathore8492, Sorry for the late reply. We filtered project list based on user access, if a user is not invited to a project won't be part of the project list. 

Commit reference - https://github.com/nocodb/nocodb/pull/4426/commits/2fe1b18b2c4b3ab22620394da47ecc6b33e14855

**Comment by pranavxc** [comment]
Hi, @vijayrathore8492 sorry for the delayed reply. It will be helpful if you can do a rebase. We made a few changes related to the project list so you may have to update `Project.list`(for getting all projects, used for super admin) and `ProjectUser.getProjectsList`(for getting invited project list).  For super admin, you can show all actions regardless of project level role.

**Comment by vijayrathore8492** [comment]
@wingkwong Does this PR align with the new changes? I just checked that a user can only see the projects that it has access to, other projects are not visible to him anymore.

**Comment by pranavxc** [comment]
@vijayrathore8492 : then we can close this PR.

**Comment by dstala** [comment]
@vijayrathore8492 closing this PR based on discussion above.

**Pull request closed without merging**